### ISSUE TITLE
Why isn't it possible to install multiple instances of a given distro without using hacks/workarounds?

### ISSUE THREAD

**Issue by abalter** [microsoft/WSL#9977]
### Windows Version

Microsoft Windows [Version 10.0.22621.1555]

### WSL Version

1.2.0.0

### Are you using WSL 1 or WSL 2?

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

5.15.90.1

### Distro Version

Ubuntu-22.04

### Other Software

None.

### Repro Steps

wsl --install -d Ubuntu-22.04

### Expected Behavior

I would like to understand why it is not possible to easily install multiple instances of a given Ubuntu version. There are hacks and workarounds. Why isn't it just normal?

### Actual Behavior

C:\Users\ariel>wsl --install -d Ubuntu-22.04
Ubuntu 22.04 LTS is already installed.
Launching Ubuntu 22.04 LTS...

### Diagnostic Logs

C:\Users\ariel>wsl -v
WSL version: 1.2.0.0
Kernel version: 5.15.90.1
WSLg version: 1.0.51
MSRDC version: 1.2.3770
Direct3D version: 1.608.2-61064218
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.22621.1555

C:\Users\ariel>wsl --install -d Ubuntu-22.04
Ubuntu 22.04 LTS is already installed.
Launching Ubuntu 22.04 LTS...

**Comment by Biswa96** [comment]
> I would like to understand why it is not possible to easily install multiple instances of a given Ubuntu version.

It is possible to install same distribution with different name. The only catch is the different registered name. Otherwise, how would any user know which one is being used? The required command is `wsl --import` and it is well documented.

**Comment by abalter** [comment]
This is the help documentation:

```
    --install [Distro] [Options...]
        Install a Windows Subsystem for Linux distribution.
        For a list of valid distributions, use 'wsl.exe --list --online'.

        Options:
            --no-launch, -n
                Do not launch the distribution after install.

            --web-download
                Download the distribution from the internet instead of the Microsoft Store.

```

It does not say anything about being able to register the distribution under a given name. The only _documented_ way is to export the distro and re-import it under a different name.

**Comment by Biswa96** [comment]
> It does not say anything about being able to register the distribution under a given name

I agree that it is not straight-forward as `wsl --install` command. That command selects the installer from this list https://github.com/microsoft/WSL/blob/master/distributions/DistributionInfo.json. The distribution name is hard coded in the installer.

> The only documented way is to export the distro and re-import it under a different name.

That is also a valid way. There are different ways that can be achieved. For example, the Ubuntu rootfs tarball can be downloaded from here https://cloud-images.ubuntu.com/wsl/ and install it using `wsl --import` command with different name.

**Comment by abalter** [comment]
>There are different ways that can be achieved
Yes. That is exactly my point though.

I guess I need to re-file this as a feature request for an option in `wsl --install` to assign a name.

**Comment by Biswa96** [comment]
> I guess I need to re-file this as a feature request for an option in wsl --install to assign a name.

Yeah, that would be another way also. But adding two names in one `wsl --install` command may confuse users.

**Comment by abalter** [comment]
How do you mean? I'm thinking of:

`wsl --install <distro> [<user supplied name>]`

So

`wsl --install Ubuntu-22.04 ubuntu1`
`wsl --install Ubuntu-22.04 ubuntu2`

would give

```
C:\>wsl --list --all
Windows Subsystem for Linux Distributions:
ubuntu1 (Default)
ubuntu2
```

If I want to know what distro they came from, I could always do

```
C:\>wsl -d ubuntu2 lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.2 LTS
Release:        22.04
Codename:       jammy
```

**Comment by Biswa96** [comment]
> wsl --install <distro> [<user supplied name>]

Good idea. There is a chance `<distro>` and `<user supplied name>` get confused. How about the command below?

```
wsl --install <distro> --name <user supplied name>
```

Inspired by `docker run <image> --name <container>` command.

**Comment by abalter** [comment]
Sure. I don't love the wsl interface anyway, and that would be an improvement.

**Comment by OneBlue** [comment]
> > `wsl --install <distro> [<user supplied name>]`
> 
> Good idea. There is a chance `<distro>` and `<user supplied name>` get confused. How about the command below?
> 
> ```
> wsl --install <distro> --name <user supplied name>
> ```
> 
> Inspired by `docker run <image> --name <container>` command.

This is a good suggestion but it's unfortunately not possible with the current model because wsl --install calls the underlying package (ubuntu.exe, debian.exe) and such.

This third party executable is what actually installs the distro, via [WslRegisterDistribution](https://learn.microsoft.com/en-us/windows/win32/api/wslapi/nf-wslapi-wslregisterdistribution), passing the name parameter (which is hardcoded in each application).

So to summarize, wsl doesn't actually control the names of the distributions being installed via wsl --instal, the distribution applications do.

To have finer control over the distribution's names, the best way would be to use wsl --import

**Comment by OneBlue** [comment]
Closing since this feature is accessible via wsl --import

**Comment by abalter** [comment]
Is there a repository somewhere of pre-made importable base images equivalent to, for instance, Ubuntu-22.04?

**Comment by Biswa96** [comment]
The WSL packages list is here https://github.com/microsoft/WSL/blob/master/distributions/DistributionInfo.json. Alternatively, each Linux distributions may have their own website to provide rootfs tarball. For example, Ubuntu has https://cloud-images.ubuntu.com/wsl/

**Comment by NotTheDr01ds** [comment]
@abalter Short aside, before I answer your most recent question -- I agree that it would be "nice to have" a `wsl --install <distro> -name <name>` version, but also understand OneBlue's point about [distro-launcher](https://github.com/microsoft/WSL-DistroLauncher) precluding that.

I see two options here.  See the bottom of this post for how I handle this, but first ...

> Is there a repository somewhere of pre-made importable base images equivalent to, for instance, Ubuntu-22.04?

Once you've installed one distribution of a given type from the Store, the pristine rootfs for it is still kept around until you uninstall the *app*, so you can `wsl --import` it.

* Open an Admin PowerShell to find the path to the rootfs:

    ```
    Get-ChildItem -Recurse 'C:\Program Files\WindowsApps\' | Where-Object {$_.Name -eq 'install.tar.gz' }

    Mode                 LastWriteTime         Length Name
    ----                 -------------         ------ ----
    -a---          2022-04-18  6:32 PM      512935559 install.tar.gz

        Directory: C:\Program Files\WindowsApps\CanonicalGroupLimited.UbuntuPreview_2210.0.148.0_x64__79rhkp1fndgsc

    Mode                 LastWriteTime         Length Name
    ----                 -------------         ------ ----
    -a---          2022-11-25  6:10 PM      446905872 install.tar.gz

        Directory: C:\Program Files\WindowsApps\CanonicalGroupLimited.Ubuntu_2204.1.8.0_x64__79rhkp1fndgsc

    Mode                 LastWriteTime         Length Name
    ----                 -------------         ------ ----
    -a---          2023-01-19  9:53 AM      562944255 install.tar.gz

        Directory: C:\Program Files\WindowsApps\TheDebianProject.DebianGNULinux_1.14.0.0_x64__76v4gfsz19hv4
    ```

* From a non-admin PowerShell, `wsl --import` the `install.tar.gz` from the given path above.  So, for example, if you wanted to create a new Ubuntu-22.04, you could:

    ```
    mkdir -p d:\WSL\instances\ubuntu_testing
    wsl --import ubuntu_testing d:\WSL\instances\ubuntu_testing "C:\Program Files\WindowsApps\CanonicalGroupLimited.Ubuntu_2204.1.8.0_x64__79rhkp1fndgsc\install.tar.gz"
    ```

    Remember to quote the `Program Files` path that has spaces.

* Of course, whether you `--import` it from a rootfs that you downloaded or the one already on your system, there are differences between this and a distribution installed with `ubuntu2204.exe` (the Distro Launcher version).  Primarily, the end-user account/password isn't created and isn't set to "default".

    So if you use this method, you'll need to manually create your non-root user, assign it to the appropriate groups (especially the `sudo` group), and assign a password.  For Ubuntu 22.04, the "normal" groups that are assigned for your user are `ntd adm dialout cdrom floppy sudo audio dip video plugdev netdev`.  For full details (using 20.04, at least), see [my Ask Ubuntu answer](https://askubuntu.com/a/1360294/1165986) on the topic.

    You'll also need to set it as default using the `/etc/wsl.conf` as I mention in that answer.

---

### Alternative Method - "Rename" the first distribution, then use `ubuntu2204.exe` again

I keep multiple distributions of the same type around myself.  My daily-driver, for instance, is `ntd_ubuntu_22.04_WSL2`.  Then, if I need to spin up a temporary one to test something in without polluting the main, I can simply `ubuntu2204.exe` again.  About 15 seconds later, I have a new distro.

As you are obviously aware, you can create as many `--import`ed distributions as you want.  So if I want a new Ubuntu 22.04 in which to try out StableLM, I can easily do that by first creating the distro above (`ubuntu2204.exe`), then:

```
# Get the location of the installed Ubuntu-22.04:
$Ubuntu2204_Path = (Get-ChildItem HKCU:\Software\Microsoft\Windows\CurrentVersion\Lxss\ |
     ForEach-Object {
         (Get-ItemProperty $_.PSPATH) | Select-Object DistributionName,BasePath
     } | Where-Object { $_.DistributionName -eq "Ubuntu-22.04" }).BasePath

# wsl --terminate or --shutdown if needed to release the lock on the existing vhdx

# Copy/import it to a new distribution
mkdir -p d:\wsl\instances\StableLM
wsl --import --vhd StableLM D:\wsl\instances\StableLM-Ubuntu22.04 "$Ubuntu2204_Path\ext4.vhdx"
```

Of course, set the default username via `/etc/wsl.conf` as mentioned above.

Then you can `wsl --unregister Ubuntu-22.04` (assuming it was just created and now copied, so that you are ready to use it again (via `ubuntu2204.exe`) the next time you need a new or temp distro based on it.

**Comment by jordantgh** [comment]
@NotTheDr01ds By far the simplest and most explicit explanation of this I've yet seen. Thank you. Are you aware of any little hacks to make launching into other distros easier? I know it sounds silly, but I like the convenience of the built in "Ubuntu" app in Windows launching me into a terminal as is the case for my main WSL2 distro.

**Comment by HendersonSC** [comment]
@jordantgh I think the first half of what you want is `wsl -d DistroName --cd ~`. The second half is to create a new shortcut with something to the effect of `powershell.exe -Command "wsl.exe -d DistroName --cd ~` in the target section. 

Windows Terminal might also provide you with options you like depending on your preferences.

**Comment by nick4fake** [comment]
@OneBlue I am sorry, but are you serious? Instead of providing the most obvious and straighforward way to specify distro name (--name) you suggest exporting/importing a distribution to... specify a name?

**Comment by NotTheDr01ds** [comment]
@jordantgh So sorry for missing your reply earlier.

I use Windows Terminal + PowerToys Run (from [Microsoft PowerToys](https://learn.microsoft.com/en-us/windows/powertoys/), installable from the Microsoft Store).

Pressing <kbd>Alt</kbd>+<kbd>SPACE</kbd> launches PowerToys Run, then typing `_` into the search box will filter on Windows Terminal Profiles.  So `_Ub` will bring up list of all of my Ubuntu profiles, with the most recently used being first.

**Comment by copdips** [comment]
1. check your existing distro by `wsl -l`
2. use `wsl --list --online` to get the list of online available distribution names
3. then install additional distribution by `wsl --install <online distro name> --name <local distro name different than the existing one>`

for e.g.:  `wsl --install Ubuntu-24.04 --name my-new-ubuntu-24.04 --location D:\wsl-distros\Ubuntu-24.04-LTS-test`

**Comment by Roemer** [comment]
> 1. check your existing distro by `wsl -l`
> 2. use `wsl --list --online` to get the list of online available distribution names
> 3. then install additional distribution by `wsl --install <online distro name> --name <local distro name different than the existing one>`
> 
> for e.g.: `wsl --install Ubuntu-24.04 --name my-new-ubuntu-24.04 --location D:\wsl-distros\my-new-ubuntu-24.04`

Since when does this work? But it does and it is great!

**Comment by B1LLP4RK** [comment]
@copdips Thank you so much! is the `--location` option necessary?
